### PR TITLE
Port `ff` crate

### PR DESCRIPTION
**Pull request by winston-h-zhang**

### CODE REVIEW

File: FF/Lib.lean [hunk 1]
```diff
@@ -0,0 +1,147 @@
+import FF.BitVec
+
+namespace FF
+
+class Field (K : Type _) extends Neg K, Add K, Sub K, Mul K where
+  /-- The zero element of the field, the additive identity. -/
+  zero : K
+  /-- The one element of the field, the multiplicative identity. -/
+  one : K
+  /-- Squares this element. -/
```
Comment (DanielRrr) [inline FF/Lib.lean:10]: Let's add `extends Ring K` and remove `zero` and `one`

File: FF/Lib.lean [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+  square : K → K
+  /-- Doubles this element. -/
+  double : K → K
+  /-- Computes the multiplicative inverse of this element.
+    Following mathlib convention, we do not fail on 0.
+    
+    TODO: Maybe should be `K → Option K` -/
+  inv : K → K
```
Comment (DanielRrr) [inline FF/Lib.lean:18]: Every element in a field has inverse, so it must be `K -> K`

File: FF/Lib.lean [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+  /-- How many bits of information can be reliably stored in the field element.
+
+    This is usually `NUM_BITS - 1`. -/
+  CAPACITY : Nat
+  /-- Inverse of 2 in the field. -/
+  TWO_INV : K
+  /-- A fixed multiplicative generator of `modulus - 1` order. 
+    This element must also be a quadratic nonresidue. -/
+  MULTIPLICATIVE_GENERATOR : K
+  /-- An integer `s` satisfying the equation `2^s * t = modulus - 1` with `t` odd.
```
Comment (DanielRrr) [inline FF/Lib.lean:78]: @mpenciak 

Do we need multiplicative generators anywere?

File: FF/Lib.lean [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    This is the number of leading zero bits in the little-endian bit representation of `modulus - 1`. -/
+  S : Nat
+  /-- The `2^s` root of unity. 
+  
+    It can be calculated by exponentiating `MULTIPLICATIVE_GENERATOR` by `t`, 
+    where `t = (modulus - 1) >> S`. 
+    
+    TODO: Enforce with proof? -/
+  ROOT_OF_UNITY : K
+  /-- Inverse of `ROOT_OF_UNITY`.
```
Comment (DanielRrr) [inline FF/Lib.lean:89]: @mpenciak 

The same question about roots of unity. I've introduced roots of unity once, but we decided to remove it.

File: FF/Lib.lean [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+    
+    TODO: Enforce with proof? -/
+  ROOT_OF_UNITY : K
+  /-- Inverse of `ROOT_OF_UNITY`. 
+  
+    TODO: Enforce with proof? -/
+  ROOT_OF_UNITY_INV : K
+  /-- Generator of the `t-order` multiplicative subgroup. -/
+  DELTA : K
+
```
Comment (DanielRrr) [inline FF/Lib.lean:95]: I think this class lacks the method that would reflect the field characteristic.

File: FF/Lib.lean [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+  TODO: IDK what I'm doing here?? -/
+class WithSmallOrderMulGroup (K : Type _) (Repr : outParam $ Type _) extends PrimeField K Repr where
+  /-- A field element of small multiplicative order `N`. -/
+  ZETA : K
+
+/-- This represents the bits of an element of a prime field. -/
+class PrimeFieldBits (K : Type _) (Repr : outParam $ Type _) extends PrimeField K Repr where
+  /-- Converts an element of the prime field into a little-endian sequence of bits. -/
+  toLEBits : K → BitArray
+  /-- Returns the bits of the field characteristic (the modulus) in little-endian order. -/
```
Comment (DanielRrr) [inline FF/Lib.lean:144]: We need conversion to both `BitArray`s and `ByteArray`s?

File: FF/Lib.lean (hunk 2 above)
Comment (winston-h-zhang) [reply FF/Lib.lean:18]: 0 doesn't have an inverse

**Pull request merged**